{"cells":[{"cell_type":"markdown","id":"3f53b037","metadata":{"id":"3f53b037"},"source":["## 1.Transcript Selection"]},{"cell_type":"code","execution_count":null,"id":"19d11fbb","metadata":{"id":"19d11fbb"},"outputs":[],"source":["import pandas as pd\n","import numpy as np"]},{"cell_type":"markdown","id":"11d58764","metadata":{"id":"11d58764"},"source":["### 1.1 Transcript Summary\n","* The Top10Gene_exon.csv is generated by Jiaqi in Gene Prioritization Process.\n","* The V2.x and V3.x and transcript_start is used for locating SNP.\n","* The V6.x is the wt nucleotide and the V7.x is the SNP"]},{"cell_type":"code","execution_count":null,"id":"bd52220d","metadata":{"id":"bd52220d","outputId":"e0847878-f31d-4b19-cccb-0462389476af"},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>V1.x</th>\n","      <th>V2.x</th>\n","      <th>V3.x</th>\n","      <th>V6.x</th>\n","      <th>V7.x</th>\n","      <th>gene_id</th>\n","      <th>transcript_id</th>\n","      <th>transcript_start</th>\n","      <th>transcript_end</th>\n","      <th>count</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr20</td>\n","      <td>5280261</td>\n","      <td>5280262</td>\n","      <td>G</td>\n","      <td>T</td>\n","      <td>ENSG00000101292.8_9</td>\n","      <td>ENST00000217270.4_5</td>\n","      <td>5279906</td>\n","      <td>5297060</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr20</td>\n","      <td>5281326</td>\n","      <td>5281327</td>\n","      <td>G</td>\n","      <td>A</td>\n","      <td>ENSG00000101292.8_9</td>\n","      <td>ENST00000217270.4_5</td>\n","      <td>5279906</td>\n","      <td>5297060</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr20</td>\n","      <td>5281252</td>\n","      <td>5281253</td>\n","      <td>G</td>\n","      <td>A</td>\n","      <td>ENSG00000101292.8_9</td>\n","      <td>ENST00000217270.4_5</td>\n","      <td>5279906</td>\n","      <td>5297060</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr20</td>\n","      <td>5280332</td>\n","      <td>5280333</td>\n","      <td>T</td>\n","      <td>C</td>\n","      <td>ENSG00000101292.8_9</td>\n","      <td>ENST00000217270.4_5</td>\n","      <td>5279906</td>\n","      <td>5297060</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr20</td>\n","      <td>5281412</td>\n","      <td>5281413</td>\n","      <td>C</td>\n","      <td>T</td>\n","      <td>ENSG00000101292.8_9</td>\n","      <td>ENST00000217270.4_5</td>\n","      <td>5279906</td>\n","      <td>5297060</td>\n","      <td>1</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["    V1.x     V2.x     V3.x V6.x V7.x              gene_id  \\\n","0  chr20  5280261  5280262    G    T  ENSG00000101292.8_9   \n","1  chr20  5281326  5281327    G    A  ENSG00000101292.8_9   \n","2  chr20  5281252  5281253    G    A  ENSG00000101292.8_9   \n","3  chr20  5280332  5280333    T    C  ENSG00000101292.8_9   \n","4  chr20  5281412  5281413    C    T  ENSG00000101292.8_9   \n","\n","         transcript_id  transcript_start  transcript_end  count  \n","0  ENST00000217270.4_5           5279906         5297060      1  \n","1  ENST00000217270.4_5           5279906         5297060      1  \n","2  ENST00000217270.4_5           5279906         5297060      1  \n","3  ENST00000217270.4_5           5279906         5297060      1  \n","4  ENST00000217270.4_5           5279906         5297060      1  "]},"execution_count":2,"metadata":{},"output_type":"execute_result"}],"source":["data=pd.read_csv(\"Top10genes_exon.csv\")\n","data[\"count\"]=[1]*data.shape[0]\n","data.head()"]},{"cell_type":"code","execution_count":null,"id":"babf8759","metadata":{"id":"babf8759","outputId":"45e3c47e-f586-46d0-ab0a-d71ccea5e49e"},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th></th>\n","      <th>count</th>\n","    </tr>\n","    <tr>\n","      <th>gene_id</th>\n","      <th>transcript_id</th>\n","      <th></th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>ENSG00000196090.13_14</th>\n","      <th>ENST00000356100.6_8</th>\n","      <td>14</td>\n","    </tr>\n","    <tr>\n","      <th>ENSG00000149633.12_15</th>\n","      <th>ENST00000279024.9_10</th>\n","      <td>14</td>\n","    </tr>\n","    <tr>\n","      <th>ENSG00000149596.7_7</th>\n","      <th>ENST00000372980.4_6</th>\n","      <td>25</td>\n","    </tr>\n","    <tr>\n","      <th>ENSG00000149488.14_7</th>\n","      <th>ENST00000358864.2_3</th>\n","      <td>17</td>\n","    </tr>\n","    <tr>\n","      <th>ENSG00000130702.15_9</th>\n","      <th>ENST00000252999.7_5</th>\n","      <td>30</td>\n","    </tr>\n","    <tr>\n","      <th rowspan=\"2\" valign=\"top\">ENSG00000130589.16_9</th>\n","      <th>ENST00000467148.2_7</th>\n","      <td>26</td>\n","    </tr>\n","    <tr>\n","      <th>ENST00000427522.6_4</th>\n","      <td>22</td>\n","    </tr>\n","    <tr>\n","      <th rowspan=\"3\" valign=\"top\">ENSG00000101292.8_9</th>\n","      <th>ENST00000217270.4_5</th>\n","      <td>18</td>\n","    </tr>\n","    <tr>\n","      <th>ENST00000678059.1_2</th>\n","      <td>18</td>\n","    </tr>\n","    <tr>\n","      <th>ENST00000678254.1_4</th>\n","      <td>18</td>\n","    </tr>\n","    <tr>\n","      <th>ENSG00000101276.18_16</th>\n","      <th>ENST00000675066.1_2</th>\n","      <td>15</td>\n","    </tr>\n","    <tr>\n","      <th>ENSG00000101181.18_9</th>\n","      <th>ENST00000370823.8_3</th>\n","      <td>15</td>\n","    </tr>\n","    <tr>\n","      <th>ENSG00000101160.15_9</th>\n","      <th>ENST00000681457.1_1</th>\n","      <td>18</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["                                            count\n","gene_id               transcript_id              \n","ENSG00000196090.13_14 ENST00000356100.6_8      14\n","ENSG00000149633.12_15 ENST00000279024.9_10     14\n","ENSG00000149596.7_7   ENST00000372980.4_6      25\n","ENSG00000149488.14_7  ENST00000358864.2_3      17\n","ENSG00000130702.15_9  ENST00000252999.7_5      30\n","ENSG00000130589.16_9  ENST00000467148.2_7      26\n","                      ENST00000427522.6_4      22\n","ENSG00000101292.8_9   ENST00000217270.4_5      18\n","                      ENST00000678059.1_2      18\n","                      ENST00000678254.1_4      18\n","ENSG00000101276.18_16 ENST00000675066.1_2      15\n","ENSG00000101181.18_9  ENST00000370823.8_3      15\n","ENSG00000101160.15_9  ENST00000681457.1_1      18"]},"execution_count":3,"metadata":{},"output_type":"execute_result"}],"source":["summary=pd.pivot_table(data, values='count', index=['gene_id','transcript_id'],aggfunc=np.sum)\n","summary.sort_values(by=[\"gene_id\",'count',], ascending=False)"]},{"cell_type":"code","execution_count":null,"id":"fb7a065f","metadata":{"id":"fb7a065f"},"outputs":[],"source":["#summary.to_csv(\"summary.csv\")"]},{"cell_type":"markdown","id":"13da52e1","metadata":{"id":"13da52e1"},"source":["### 1.2 Transcript selection based on PDB data\n","* The Protein.csv is manually created by Keyi \n","* The first three column is the accession id, the column pdb and alphafold contains the information about whether the sturcture information is adopted by Ensembl. The protein_length_match column contains information about whether the downloaded structure has the same length as the wt sequence (in later part, it is furthur verified to see if the sequence of the structure matches with the wt amino acid sequence"]},{"cell_type":"code","execution_count":null,"id":"2dbb52c8","metadata":{"id":"2dbb52c8","outputId":"c8efb2e2-d519-4053-94ca-def237bd773e"},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>gene_id</th>\n","      <th>transcript_id</th>\n","      <th>uniprot</th>\n","      <th>pdb</th>\n","      <th>alphafold</th>\n","      <th>mutation</th>\n","      <th>protein_length</th>\n","      <th>Name</th>\n","      <th>Unnamed: 8</th>\n","      <th>protein_length_match</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>ENSG00000101160.15_9</td>\n","      <td>ENST00000681457.1_1</td>\n","      <td>no</td>\n","      <td>Retained intron</td>\n","      <td>no</td>\n","      <td>18</td>\n","      <td>0</td>\n","      <td>NaN</td>\n","      <td>1</td>\n","      <td>no</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>ENSG00000101181.18_9</td>\n","      <td>ENST00000370823.8_3</td>\n","      <td>Q9H4K7-1</td>\n","      <td>no</td>\n","      <td>yes</td>\n","      <td>15</td>\n","      <td>406</td>\n","      <td>MTG2-201</td>\n","      <td>2</td>\n","      <td>yes</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>ENSG00000101276.18_16</td>\n","      <td>ENST00000675066.1_2</td>\n","      <td>A0A6Q8PFQ2</td>\n","      <td>no</td>\n","      <td>no</td>\n","      <td>15</td>\n","      <td>364</td>\n","      <td>SLC52A3-207</td>\n","      <td>3</td>\n","      <td>yes</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>ENSG00000101292.8_9</td>\n","      <td>ENST00000678059.1_2</td>\n","      <td>A0A7I2V3D2</td>\n","      <td>no</td>\n","      <td>no</td>\n","      <td>18</td>\n","      <td>348</td>\n","      <td>PROKR2-202</td>\n","      <td>4</td>\n","      <td>yes</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>ENSG00000130589.16_9</td>\n","      <td>ENST00000427522.6_4</td>\n","      <td>Q9BYK8-2</td>\n","      <td>no</td>\n","      <td>no</td>\n","      <td>22</td>\n","      <td>2080</td>\n","      <td>HELZ2-202</td>\n","      <td>5</td>\n","      <td>no</td>\n","    </tr>\n","    <tr>\n","      <th>5</th>\n","      <td>ENSG00000130702.15_9</td>\n","      <td>ENST00000252999.7_5</td>\n","      <td>O15230-1</td>\n","      <td>no</td>\n","      <td>no</td>\n","      <td>30</td>\n","      <td>3695</td>\n","      <td>LAMA5-201</td>\n","      <td>6</td>\n","      <td>no</td>\n","    </tr>\n","    <tr>\n","      <th>6</th>\n","      <td>ENSG00000149488.14_7</td>\n","      <td>ENST00000358864.2_3</td>\n","      <td>Q8TDI7-1</td>\n","      <td>no</td>\n","      <td>yes</td>\n","      <td>17</td>\n","      <td>906</td>\n","      <td>TMC2-201</td>\n","      <td>7</td>\n","      <td>no</td>\n","    </tr>\n","    <tr>\n","      <th>7</th>\n","      <td>ENSG00000149596.7_7</td>\n","      <td>ENST00000372980.4_6</td>\n","      <td>Q9BR39-1</td>\n","      <td>no</td>\n","      <td>yes</td>\n","      <td>25</td>\n","      <td>696</td>\n","      <td>JPH2-202</td>\n","      <td>8</td>\n","      <td>yes</td>\n","    </tr>\n","    <tr>\n","      <th>8</th>\n","      <td>ENSG00000149633.12_15</td>\n","      <td>ENST00000279024.9_10</td>\n","      <td>Q5JYT7</td>\n","      <td>no</td>\n","      <td>yes</td>\n","      <td>14</td>\n","      <td>1200</td>\n","      <td>KIAA1755-201</td>\n","      <td>9</td>\n","      <td>yes</td>\n","    </tr>\n","    <tr>\n","      <th>9</th>\n","      <td>ENSG00000196090.13_14</td>\n","      <td>ENST00000356100.6_8</td>\n","      <td>B1AJR8</td>\n","      <td>no</td>\n","      <td>no</td>\n","      <td>14</td>\n","      <td>1450</td>\n","      <td>PTPRT-201</td>\n","      <td>10</td>\n","      <td>no</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["                 gene_id         transcript_id     uniprot              pdb  \\\n","0   ENSG00000101160.15_9   ENST00000681457.1_1          no  Retained intron   \n","1   ENSG00000101181.18_9   ENST00000370823.8_3    Q9H4K7-1               no   \n","2  ENSG00000101276.18_16   ENST00000675066.1_2  A0A6Q8PFQ2               no   \n","3    ENSG00000101292.8_9   ENST00000678059.1_2  A0A7I2V3D2               no   \n","4   ENSG00000130589.16_9   ENST00000427522.6_4    Q9BYK8-2               no   \n","5   ENSG00000130702.15_9   ENST00000252999.7_5    O15230-1               no   \n","6   ENSG00000149488.14_7   ENST00000358864.2_3    Q8TDI7-1               no   \n","7    ENSG00000149596.7_7   ENST00000372980.4_6    Q9BR39-1               no   \n","8  ENSG00000149633.12_15  ENST00000279024.9_10      Q5JYT7               no   \n","9  ENSG00000196090.13_14   ENST00000356100.6_8      B1AJR8               no   \n","\n","  alphafold  mutation  protein_length          Name  Unnamed: 8  \\\n","0        no        18               0           NaN           1   \n","1       yes        15             406      MTG2-201           2   \n","2        no        15             364   SLC52A3-207           3   \n","3        no        18             348    PROKR2-202           4   \n","4        no        22            2080     HELZ2-202           5   \n","5        no        30            3695     LAMA5-201           6   \n","6       yes        17             906      TMC2-201           7   \n","7       yes        25             696      JPH2-202           8   \n","8       yes        14            1200  KIAA1755-201           9   \n","9        no        14            1450     PTPRT-201          10   \n","\n","  protein_length_match  \n","0                   no  \n","1                  yes  \n","2                  yes  \n","3                  yes  \n","4                   no  \n","5                   no  \n","6                   no  \n","7                  yes  \n","8                  yes  \n","9                   no  "]},"execution_count":4,"metadata":{},"output_type":"execute_result"}],"source":["protein=pd.read_csv(\"Protein.csv\")\n","protein"]},{"cell_type":"markdown","id":"2752795f","metadata":{"id":"2752795f"},"source":["## 2. Data Preprocessing"]},{"cell_type":"markdown","id":"8a92c7ea","metadata":{"id":"8a92c7ea"},"source":["### 2.1 Read in all sequence\n","* Each transcript has a folder named by its corresponding index. In each folder, there are three fa files downloaded from Ensembl(https://useast.ensembl.org/index.html) and a pdb.fa file downloaded from the UniProt. genomics.fa is downloaded at the transcript page, exon panel by selecting download genomic sequence with 0 flanking sequence. exon.fa is downloaded at the transcript page, cDNA panel by selecing only exon. The amino.fa is downloaded at the transcript page, protein panel. The table.csv contains information about exon and intron and is downloaded at the transcript page and exon panel.\n","* The structure in each folder is downloaded from alphafold database (https://alphafold.ebi.ac.uk/) using accession number. Only structure with exactly the same sequence as WT are downloaded and analyzed"]},{"cell_type":"code","execution_count":null,"id":"33c631e3","metadata":{"id":"33c631e3","outputId":"6ec35654-3c2e-4ead-8fc8-38cfe04f580f"},"outputs":[{"data":{"text/plain":["['9/', '8/', '4/', '3/', '2/']"]},"execution_count":8,"metadata":{},"output_type":"execute_result"}],"source":["import os\n","import string\n","folders=[]\n","for folder in os.listdir(\"data/\"):\n","    if \".\" not in folder:\n","        folder=folder+\"/\"\n","        folders.append(folder)\n","folders"]},{"cell_type":"code","execution_count":null,"id":"30899d04","metadata":{"id":"30899d04"},"outputs":[],"source":["# save all sequences\n","all_sequence={}\n","for folder in folders:\n","    all_sequence[folder[:-1]]={}\n","    for file in os.listdir(\"data/\"+folder):\n","        if \"fa\" in file:\n","            f=open(\"data/\"+folder+file)\n","            ls=[]\n","            for line in f:\n","                if not line.startswith('>'):    \n","                    ls.append(line.replace('\\n',''))\n","            sequence=\"\".join(ls)\n","            all_sequence[folder[:-1]][file.split(\".\")[0]]=sequence\n","            f.close()"]},{"cell_type":"code","execution_count":null,"id":"9d5afcdf","metadata":{"id":"9d5afcdf","outputId":"9bfad652-8da5-4031-cfdf-15ebf7bd58ac"},"outputs":[{"data":{"text/plain":["{'9': 'ENST00000279024.9_10',\n"," '8': 'ENST00000372980.4_6',\n"," '4': 'ENST00000678059.1_2',\n"," '3': 'ENST00000675066.1_2',\n"," '2': 'ENST00000370823.8_3'}"]},"execution_count":19,"metadata":{},"output_type":"execute_result"}],"source":["# save corresponding transcrip_id\n","transcript_ids = {}\n","for item in all_sequence.keys():\n","    index=int(item)-1\n","    transcript_ids[item]=protein[\"transcript_id\"].to_list()[index]\n","transcript_ids"]},{"cell_type":"markdown","id":"907f8cf9","metadata":{"id":"907f8cf9"},"source":["### 2.2 Check PDB sequence and Ensembl Amino sequence"]},{"cell_type":"code","execution_count":null,"id":"ec22c6a5","metadata":{"id":"ec22c6a5","outputId":"b54c6d9f-e65b-4f07-ae82-1c30b65fab96"},"outputs":[{"data":{"text/plain":["dict_keys(['9', '8', '4', '3', '2'])"]},"execution_count":15,"metadata":{},"output_type":"execute_result"}],"source":["all_sequence.keys()"]},{"cell_type":"code","execution_count":null,"id":"4aaa3cdc","metadata":{"id":"4aaa3cdc","outputId":"c67e7d80-a353-44ef-c2b4-3c2cad2dc07a"},"outputs":[{"data":{"text/plain":["dict_keys(['genomic', 'amino', 'pdb', 'exon'])"]},"execution_count":16,"metadata":{},"output_type":"execute_result"}],"source":["all_sequence[\"9\"].keys()"]},{"cell_type":"code","execution_count":null,"id":"4709c4ab","metadata":{"id":"4709c4ab","outputId":"26d152bb-8cde-4762-8c19-47dc32d97bf8"},"outputs":[{"name":"stdout","output_type":"stream","text":["True: 9\n","True: 8\n","True: 4\n","True: 3\n","True: 2\n"]}],"source":["# check if the wt sequence from the ensemble matches the sequence of the corresponding structure in alphafold database\n","for item in all_sequence.keys():\n","    if all_sequence[item][\"pdb\"]==all_sequence[item][\"amino\"]:\n","        print(\"True: \"+item)"]},{"cell_type":"markdown","id":"d018e961","metadata":{"id":"d018e961"},"source":["### 2.3 Create Mutant Sequence and Save"]},{"cell_type":"code","execution_count":null,"id":"a9b59404","metadata":{"id":"a9b59404"},"outputs":[],"source":["# table for splicing info\n","def table_info(filename):\n","    data=pd.read_csv(filename,thousands=',').iloc[1:-1,:]\n","    data[[\"Start\",\"End\",\"Length\"]]=data[[\"Start\",\"End\",\"Length\"]].astype(\"int\")\n","    data=data[data[\"Exon / Intron\"].apply(lambda x: x.startswith(\"E\"))]\n","    return data"]},{"cell_type":"code","execution_count":null,"id":"3daa2d24","metadata":{"id":"3daa2d24"},"outputs":[],"source":["# Is it anti-sense\n","def anti(table):\n","    if (table[\"Start\"].to_list()[0])>(table[\"End\"].to_list()[0]):\n","        return True\n","    else:\n","        return False"]},{"cell_type":"code","execution_count":null,"id":"8fd9ee06","metadata":{"id":"8fd9ee06"},"outputs":[],"source":["# complimentary sequence\n","def compliment(sequence):\n","    comp_dict = {\n","        \"A\":\"T\",\n","        \"T\":\"A\",\n","        \"G\":\"C\",\n","        \"C\":\"G\",\n","    }\n","    sequence_list = list(sequence)\n","    sequence_list = [comp_dict[base] for base in sequence_list]\n","    string = ''.join(sequence_list)\n","    return string"]},{"cell_type":"code","execution_count":null,"id":"3f391d4b","metadata":{"id":"3f391d4b"},"outputs":[],"source":["# Create Mutant: This function can introduce mutation in to the right place through relative position.\n","# It can be sued for both sense and anti-sense cases\n","def mutants(transcript_table,id,wt_seq,anti):\n","    transcript_table=transcript_table[transcript_table[\"transcript_id\"].apply(lambda x: x.startswith(id))]\n","    if anti:\n","        wt_seq=compliment(wt_seq)\n","        wt_seq=list(wt_seq.strip())\n","        loc=(transcript_table[\"V3.x\"]-transcript_table[\"transcript_start\"]).to_list()\n","        n=0\n","        for i in range(len(loc)):\n","            item=loc[i]\n","            if (wt_seq[-item]==transcript_table[\"V6.x\"].to_list()[i])==False:\n","                n+=1\n","            wt_seq[-item]=transcript_table[\"V7.x\"].to_list()[i] \n","        print(\"Mismatch between reference and ensembl: \"+str(n))\n","        wt_seq=\"\".join(wt_seq)\n","        wt_seq=compliment(wt_seq)\n","    else:\n","        loc=(transcript_table[\"V2.x\"]-transcript_table[\"transcript_start\"]).to_list()\n","        wt_seq=list(wt_seq.strip())\n","        n=0\n","        for i in range(len(loc)):\n","            item=loc[i]\n","            if (wt_seq[item]==transcript_table[\"V6.x\"].to_list()[i])==False:\n","                n+=1\n","           # else:\n","             #   wt_seq[item]=transcript_table[\"V7.x\"].to_list()[i]\n","        print(\"Mismatch between reference and ensembl: \"+str(n))\n","        wt_seq=\"\".join(wt_seq)\n","    return wt_seq"]},{"cell_type":"code","execution_count":null,"id":"07220def","metadata":{"id":"07220def"},"outputs":[],"source":["# splice: exract exons from the given transcript sequence\n","def exons(table,wt_seq):\n","    relative_start=[]\n","    start=table[\"Start\"].to_list()[0]\n","    for item in table[\"Start\"].to_list():\n","        relative_start.append(np.abs(item-start))\n","    relative_end=[]\n","    for item in table[\"End\"].to_list():\n","        relative_end.append(np.abs(item-start)+1)\n","    final_seq=\"\"\n","    for item in zip(relative_start,relative_end):\n","        final_seq+=wt_seq[item[0]:item[1]]\n","    return final_seq"]},{"cell_type":"code","execution_count":null,"id":"034a1ddb","metadata":{"id":"034a1ddb"},"outputs":[],"source":["# exon to amino\n","def translate(seq,seq_dict): \n","    table = {\n","        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n","        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n","        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n","        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                \n","        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n","        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n","        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n","        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n","        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n","        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n","        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n","        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n","        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n","        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n","        'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n","        'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',\n","    }\n","    loc=seq.find(\"ATG\")\n","    seq=seq[loc:]\n","    protein =\"\"\n","    while (len(protein)!=len(seq_dict[\"amino\"]) and (\"ATG\" in seq)):\n","        protein=\"\"\n","        for i in range(0, len(seq), 3):\n","            codon = seq[i:i + 3]\n","            if table[codon]!=\"_\":\n","                protein+= table[codon]\n","            else:\n","                break\n","        if len(protein)!=len(seq_dict[\"amino\"]):\n","            seq=seq[1:]\n","            if \"ATG\" in seq:\n","                loc=seq.find(\"ATG\")\n","                seq=seq[loc:]\n","            else:\n","                break\n","        else:\n","            break\n","    return protein"]},{"cell_type":"code","execution_count":null,"id":"d3efb35b","metadata":{"id":"d3efb35b"},"outputs":[],"source":["# integrated function\n","def workflow(filename,transcript_table,seq_dict,id):\n","    table=table_info(filename)\n","    wt_seq=seq_dict[\"genomic\"]\n","    flag=anti(table)\n","    mutant_seq=mutants(transcript_table,id,wt_seq,flag)\n","    #print(\"mutant:\" + str(mutant_seq==seq_dict[\"genomic\"]) )\n","    exon=exons(table,mutant_seq)\n","    #print(\"Exon:\" + str(exon==seq_dict[\"exon\"]))\n","    protein=translate(exon,seq_dict)\n","    #print(\"Protein: \"+ str(protein==seq_dict[\"amino\"]))\n","    return protein"]},{"cell_type":"code","execution_count":null,"id":"b64657de","metadata":{"id":"b64657de","outputId":"de77f4ae-8376-4eb9-a3fa-8cfe86db0ab7"},"outputs":[{"name":"stdout","output_type":"stream","text":["Mismatch between reference and ensembl: 0\n","Mismatch between reference and ensembl: 0\n","Mismatch between reference and ensembl: 0\n","Mismatch between reference and ensembl: 0\n","Mismatch between reference and ensembl: 0\n"]}],"source":["protein_list={}\n","for item in all_sequence.keys():\n","    filename=\"data/\"+item+\"/table.csv\"\n","    seq_dict=all_sequence[item]\n","    id=transcript_ids[item]\n","    protein_list[item]=workflow(filename,data,seq_dict,id)"]},{"cell_type":"code","execution_count":null,"id":"8b13208d","metadata":{"id":"8b13208d","outputId":"e5512c6a-e5c5-4c2d-ad63-aa1ae8bfa8eb"},"outputs":[{"data":{"text/plain":["{'9': 'MDPPSLDTAIQHALAGLYPPFEATAPTVLGQVFRLLDSGFQGDGLSFLLDFLIPAKRLCEQVREAACAPYSHCLFLHEGWPLCLRDEVVVHLAPLNPLLLRQGDFYLQVEPQEEQSVCIMIKCLSLDLCTVDKKPVPEPAYPILFTQEWLEAINSDFEGNPLHNCLVASENGIAPVPWTKITSPEFVDDRPQVVNALCQAWGPLPLEALDLSSPQELHQASSPDNQVLPAQSLAKGKGRTYGSKYPGLIKVEQARCGEVAFRMDEVVSQDFEGDYVALLGFSQESRGESPSREAGTSSGCTSGALEEIAGTKETPLFQKILPLSEANEGPSLGNRACTNPESSEERPYNLGFRRKVNLKAPTHNSERPPQGSYMNVLEDALDCASGLRAGVSQEPAASKMQGPLGNPENMVQLRPGPRQASSPRLSPASPAAAASETKIEVKTKERNGRLPKPMPCPSRNTSSPEPPTPGLKFSFLRGQRQPSVTPEKASLQHNGPWKVLCSLYSPKPNRAKSLGKAGTTQTKTSGPATAPSPLTEEKAALPEASAGSPERGPTLEEEPPGPEPRIGALGVKVFRSRIACLPGGRDRAGRPLLLVSTTEGAWEAPWCTVSEVTKLLSYLCTIPRPEDKAKGLAVLIDARRQPPQPGLVSALQATQAQVPASIRAILFLGEKEAALQLQTLPDVQVEVLTSLKALSHHVDPSQLPAVLEGPFPYCHTEWVHFFQKLDPFLADLHQASSLLQASIEEFEKADPPGGMQEATRCLSKSKELMEAVLRDPGLLGLQREGGATLARLQHDASRLDFSPDVRSHLAAATALYSLVDEQLHVLVTASNSLLGKLELRVRLGRLEAAIHQVSDWMEQEGRRCLQSLTPKDGSLETVEKAHAEFENFFLQAAAQYRRGLELSKQAAQLGATARGAGEAERAEFPELAAFASTQRAFQAKLTHFYMAAERQRTDLETLLHLHRFCKRMTWFHMDCQDLMAQLRLDKTSRVSPGDQRRLHRYLQRLASEFPAEKLAAVGLQVASLSRAGLGQELWEEARIRHEEIWMLLEKALTHSSCPEAPAAHSARPERRGVAAKGQGVSVEVTSKGRWDQPPLDSLGMDHLPKSYWPPGPPRGEQNRTFQAGSPPQEAGQAAEAEDGKGSHKLPDPAREHLLATTFFRQQPPRQSQVPRLTGGSFSSEGTDSQTSLEDSPQTSPLASL',\n"," '8': 'MSGGRFDFDDGGAYCGGWEGGKAHGHGLCTGPKGQGEYSGSWNFGFEVAGVYTWPSGNTFEGYWSQGKRHGLGIETKGRWLYKGEWTHGFKGRYGIRQSSSSGAKYEGTWNNGLQDGYGTETYADGGTYQGQFTNGMRHGYGVRQSVPYGMAVVVRSPLRTSLSSLRSEHSNGTVAPDSPASPASDGPALPSPAIPRGGFALSLLANAEAAARAPKGGGLFQRGALLGKLRRAESRTSVGSQRSRVSFLKSDLSSGASDAASTASLGEAAEGADEAAPFEADIDATTTETYMGEWKNDKRSGFGVSERSSGLRYEGEWLDNLRHGYGCTTLPDGHREEGKYRHNVLVKDTKRRMLQLKSNKVRQKVEHSVEGAQRAAAIARQKAEIAASRTSHAKAKAEAAEQAALAANQESNIARTLARELAPDFYQPGPEYQKRRLLQEILENSESLLEPPDRGAGAAGLPQPPRESPQLHERETPRPEGGSPSPAGTPPQPKRPRPGVSKDGLLSPGAWNGEPSGEGSRSVTPSEGAGRRSPARPATERMAIEALQAPPAPSREPEVALYQGYHSYAVRTTPPEPPPFEDQPEPEVSGSESAPSSPATAPLQAPTLRGPEPARETPAKLEPKPIIPKAEPRAKARKTEARGLTKAGAKKKARKEAALAAEAEVEVEEVPNTILICMVILLNIGLAILFVHLLT',\n"," '4': 'MDEDEDMTKTRTFFAAKIVIGIALAGIMLVCGIGNFVFIAALTRYKKLRNLTNLLIANLAISDFLVAIICCPFEMDYYVVRQLSWEHGHVLCASVNYLRTVSLYVSTNALLAIAIDRYLAIVHPLKPRMNYQTASFLIALVWMVSILIAIPSAYFATETVLFIVKSQEKIFCGQIWPVDQQLYYKSYFLFIFGVEFVGPVVTMTLCYARISRELWFKAVPGFQTEQIRKRLRCRRKTVLVLMCILTAYVLCWAPFYGFTIIRDFFPTVFVKEKHYLTAFYVVECIAMSNSMINTVCFVTVKNNTMKYFKKMMLLHWRPSQRGSKSSADLDLRTNGVPTTEEVDCIRLK',\n"," '3': 'MAFLMHLLVCVFGMGSWVTINGLWVELPLLVMELPEGWYLPSYLTVVIQLANIGPLLVTLLHHFRPSCLSEVPIIFTLLGVGTVTCIIFAFLWNMTSWVLDGHHSIAFLVLTFFLALVDCTSSVTFLPFMSRLPTYYLTTFFVGEGLSGLLPALVALAQGSGLTTCVNVTEISDSVPSPVPTRETDIAQGVPRALVSALPGMEAPLSHLESRYLPAHFSPLVFFLLLSIMMACCLVAFFVLQRQPRCWEASVEDLLNDQVTLHSIRLREENDLGPAGMVDSSQGQGYLEEKAAPCCPAHLAFVYTLVAFVNALTNGMLPSVQTYSCLSYGPVAYHLAATLSIVANPLASLVSMFLPNRCALIWS',\n"," '2': 'MAPARCFSARLRTVFQGVGHWALSTWAGLKPSRLLPQRASPRLLSVGRADLAKHQELPGKKLLSEKKLKRYFVDYRRVLVCGGNGGAGASCFHSEPRKEFGGPDGGDGGNGGHVILRVDQQVKSLSSVLSRYQGFSGEDGGSKNCFGRSGAVLYIRVPVGTLVKEGGRVVADLSCVGDEYIAALGGAGGKGNRFFLANNNRAPVTCTPGQPGQQRVLHLELKTVAHAGMVGFPNAGKSSLLRAISNARPAVASYPFTTLKPHVGIVHYEGHLQIAVADIPGIIRGAHQNRGLGSAFLRHIERCRFLLFVVDLSQPEPWTQVDDLKYELEMYEKGLSARPHAIVANKIDLPEAQANLSQLRDHLGQEVIVLSALTGENLEQLLLHLKVLYDAYAEAELGQGRQPLRW'}"]},"execution_count":55,"metadata":{},"output_type":"execute_result"}],"source":["protein_list"]},{"cell_type":"code","execution_count":null,"id":"80049081","metadata":{"id":"80049081","outputId":"b4e52fa8-fe87-4222-f1ba-672e87aa9f65"},"outputs":[{"name":"stdout","output_type":"stream","text":["True\n","True\n","True\n","True\n","True\n"]}],"source":["for item in protein_list.keys():\n","    print(len(protein_list[item])==len(all_sequence[item][\"pdb\"]))"]},{"cell_type":"code","execution_count":null,"id":"141e273e","metadata":{"id":"141e273e"},"outputs":[],"source":["for item in protein_list.keys():\n","    filename=\"data/\"+item+\"/mutant_seq.txt\"\n","    with open(filename, 'w') as f:\n","        f.write(protein_list[item])"]},{"cell_type":"markdown","id":"3f58abae","metadata":{"id":"3f58abae"},"source":["### 2.4 How is mutant sequence different from WT sequence?\n","* Mutant sequence is compared to the WT sequence to see how many non-synoymous mutation exists\n","* Mutant sequence which contains non-synonymous mutations are saved and its structure is predicted using the alphafold 2 colab (https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb)"]},{"cell_type":"code","execution_count":null,"id":"ff55e95f","metadata":{"id":"ff55e95f"},"outputs":[],"source":["def dif_seq(seq1,seq2):\n","    seq1=np.array(list(seq1))\n","    seq2=np.array(list(seq2))\n","    print(\"Matches: \"+str(np.sum(seq1==seq2))+\"/\"+str(seq1.shape[0]))"]},{"cell_type":"code","execution_count":null,"id":"9af48618","metadata":{"id":"9af48618","outputId":"9de61e82-f019-4a87-b518-09a0ca7e83ee"},"outputs":[{"name":"stdout","output_type":"stream","text":["9\n","Matches: 1197/1200\n","8\n","Matches: 696/696\n","4\n","Matches: 347/348\n","3\n","Matches: 361/364\n","2\n","Matches: 406/406\n"]}],"source":["for item in protein_list.keys():\n","    print(item)\n","    seq1=protein_list[item]\n","    seq2=all_sequence[item][\"amino\"]\n","    dif_seq(seq1,seq2)"]},{"cell_type":"markdown","id":"396d1093","metadata":{"id":"396d1093"},"source":["## 3. Structure Analysis\n","* With the predicted mutant structure, LDDT is calculated to see how mutation alters the structure. Since LDDT requires exact match between the model and reference sequence. Mutation residues are removed. Calculation is performed on rank 1 sequence and wt\n","* Website for LDDT: https://swissmodel.expasy.org/lddt"]},{"cell_type":"code","execution_count":null,"id":"a4a98eeb","metadata":{"id":"a4a98eeb"},"outputs":[],"source":["def find_mismatch(seq1,seq2):\n","    seq1=np.array(list(seq1))\n","    seq2=np.array(list(seq2))\n","    flag=(seq1==seq2)\n","    print(np.where(flag==0))"]},{"cell_type":"code","execution_count":null,"id":"0d1e1249","metadata":{"id":"0d1e1249","outputId":"5443a2f5-d04b-4c5f-c855-f9e11e9ea3d7"},"outputs":[{"name":"stdout","output_type":"stream","text":["9\n","(array([ 338,  939, 1044]),)\n","8\n","(array([], dtype=int64),)\n","4\n","(array([260]),)\n","3\n","(array([266, 277, 302]),)\n","2\n","(array([], dtype=int64),)\n"]}],"source":["for item in protein_list.keys():\n","    print(item)\n","    seq1=protein_list[item]\n","    seq2=all_sequence[item][\"amino\"]\n","    find_mismatch(seq1,seq2)"]},{"cell_type":"code","execution_count":null,"id":"0ee73a40","metadata":{"id":"0ee73a40"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"eadebd2c","metadata":{"id":"eadebd2c"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"34c60d4b","metadata":{"id":"34c60d4b"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"d26330b2","metadata":{"id":"d26330b2"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"1404b826","metadata":{"id":"1404b826"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"965bee42","metadata":{"id":"965bee42"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"4d63bdc1","metadata":{"id":"4d63bdc1"},"outputs":[],"source":[]},{"cell_type":"markdown","id":"fbc72435","metadata":{"id":"fbc72435"},"source":["### Trash"]},{"cell_type":"code","execution_count":null,"id":"6797e388","metadata":{"id":"6797e388"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"2b3c4ed4","metadata":{"id":"2b3c4ed4"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"a87b1793","metadata":{"id":"a87b1793"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"17364462","metadata":{"id":"17364462"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"eaab0027","metadata":{"id":"eaab0027"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"64564c0d","metadata":{"id":"64564c0d"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"708c2a24","metadata":{"id":"708c2a24"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"3be26209","metadata":{"id":"3be26209"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"4305d8ae","metadata":{"id":"4305d8ae"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"57650f1a","metadata":{"id":"57650f1a"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"abd12591","metadata":{"id":"abd12591"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"2ab71d41","metadata":{"id":"2ab71d41"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"3847539f","metadata":{"id":"3847539f"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"4ae0cff2","metadata":{"id":"4ae0cff2"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"80139563","metadata":{"id":"80139563"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"d3becbde","metadata":{"id":"d3becbde"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"9c5d514c","metadata":{"id":"9c5d514c"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"ea444608","metadata":{"id":"ea444608"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"a40d5dd0","metadata":{"id":"a40d5dd0"},"outputs":[],"source":[]},{"cell_type":"markdown","id":"cab919e5","metadata":{"id":"cab919e5"},"source":["### 2.1 Extract wt sequence from fasta"]},{"cell_type":"code","execution_count":null,"id":"b5c5e020","metadata":{"id":"b5c5e020"},"outputs":[],"source":["def table_info(filename):\n","    data=pd.read_csv(filename,thousands=',').iloc[1:-1,:]\n","    data[[\"Start\",\"End\",\"Length\"]]=data[[\"Start\",\"End\",\"Length\"]].astype(\"int\")\n","    data=data[data[\"Exon / Intron\"].apply(lambda x: x.startswith(\"E\"))]\n","    return data\n","\n","def anti(table):\n","    if (table[\"Start\"].to_list()[0])>(table[\"End\"].to_list()[0]):\n","        return True\n","    else:\n","        return False\n","    \n","def compliment(sequence):\n","    comp_dict = {\n","        \"A\":\"T\",\n","        \"T\":\"A\",\n","        \"G\":\"C\",\n","        \"C\":\"G\",\n","    }\n","    sequence_list = list(sequence)\n","    sequence_list = [comp_dict[base] for base in sequence_list]\n","    string = ''.join(sequence_list)\n","    return string\n","\n","def mutants(transcript_table,id,wt_seq,anti):\n","    transcript_table=transcript_table[transcript_table[\"transcript_id\"].apply(lambda x: x.startswith(id))]\n","    if anti:\n","        wt_seq=compliment(wt_seq)\n","        wt_seq=list(wt_seq.strip())\n","        loc=(transcript_table[\"V3.x\"]-transcript_table[\"transcript_start\"]).to_list()\n","        n=0\n","        for i in range(len(loc)):\n","            item=loc[i]\n","            if (wt_seq[-item]==transcript_table[\"V6.x\"].to_list()[i])==False:\n","                n+=1\n","            #else:\n","               # wt_seq[-item]=transcript_table[\"V7.x\"].to_list()[i] \n","        print(\"Mismatch between reference and ensembl: \"+str(n))\n","        wt_seq=\"\".join(wt_seq)\n","        wt_seq=compliment(wt_seq)\n","    else:\n","        loc=(transcript_table[\"V2.x\"]-transcript_table[\"transcript_start\"]).to_list()\n","        wt_seq=list(wt_seq.strip())\n","        n=0\n","        for i in range(len(loc)):\n","            item=loc[i]\n","            if (wt_seq[item]==transcript_table[\"V6.x\"].to_list()[i])==False:\n","                n+=1\n","           # else:\n","             #   wt_seq[item]=transcript_table[\"V7.x\"].to_list()[i]\n","        print(\"Mismatch between reference and ensembl: \"+str(n))\n","        wt_seq=\"\".join(wt_seq)\n","    return wt_seq\n","\n","def exons(table,wt_seq):\n","    relative_start=[]\n","    start=table[\"Start\"].to_list()[0]\n","    for item in table[\"Start\"].to_list():\n","        relative_start.append(np.abs(item-start))\n","    relative_end=[]\n","    for item in table[\"End\"].to_list():\n","        relative_end.append(np.abs(item-start)+1)\n","    final_seq=\"\"\n","    for item in zip(relative_start,relative_end):\n","        final_seq+=wt_seq[item[0]:item[1]]\n","    return final_seq\n","\n","def translate(seq,seq_dict): \n","    table = {\n","        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n","        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n","        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n","        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                \n","        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n","        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n","        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n","        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n","        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n","        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n","        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n","        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n","        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n","        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n","        'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n","        'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',\n","    }\n","    loc=seq.find(\"ATG\")\n","    seq=seq[loc:]\n","    protein =\"\"\n","    while (len(protein)!=len(seq_dict[\"amino\"]) and (\"ATG\" in seq)):\n","        protein=\"\"\n","        for i in range(0, len(seq), 3):\n","            codon = seq[i:i + 3]\n","            if table[codon]!=\"_\":\n","                protein+= table[codon]\n","            else:\n","                break\n","        if len(protein)!=len(seq_dict[\"amino\"]):\n","            seq=seq[1:]\n","            if \"ATG\" in seq:\n","                loc=seq.find(\"ATG\")\n","                seq=seq[loc:]\n","            else:\n","                break\n","        else:\n","            break\n","    return protein\n","\n","def workflow(filename,transcript_table,seq_dict,id):\n","    table=table_info(filename)\n","    wt_seq=seq_dict[\"genomic\"]\n","    flag=anti(table)\n","    mutant_seq=mutants(transcript_table,id,wt_seq,flag)\n","    print(\"mutant:\" + str(mutant_seq==seq_dict[\"genomic\"]) )\n","    exon=exons(table,mutant_seq)\n","    print(\"Exon:\" + str(exon==seq_dict[\"exon_only\"]))\n","    protein=translate(exon,seq_dict)\n","    print(\"Protein: \"+ str(protein==seq_dict[\"amino\"]))\n","    return protein"]},{"cell_type":"code","execution_count":null,"id":"c86a6954","metadata":{"id":"c86a6954"},"outputs":[],"source":["def anti(table):\n","    if (table[\"Start\"].to_list()[0])>(table[\"End\"].to_list()[0]):\n","        return True\n","    else:\n","        return False"]},{"cell_type":"code","execution_count":null,"id":"50e40748","metadata":{"id":"50e40748"},"outputs":[],"source":["def compliment(sequence):\n","    comp_dict = {\n","        \"A\":\"T\",\n","        \"T\":\"A\",\n","        \"G\":\"C\",\n","        \"C\":\"G\",\n","    }\n","    sequence_list = list(sequence)\n","    sequence_list = [comp_dict[base] for base in sequence_list]\n","    string = ''.join(sequence_list)\n","    return string"]},{"cell_type":"code","execution_count":null,"id":"9218b926","metadata":{"id":"9218b926"},"outputs":[],"source":["def mutants(transcript_table,id,wt_seq,anti):\n","    transcript_table=transcript_table[transcript_table[\"transcript_id\"].apply(lambda x: x.startswith(id))]\n","    if anti:\n","        wt_seq=compliment(wt_seq)\n","        wt_seq=list(wt_seq.strip())\n","        loc=(transcript_table[\"V3.x\"]-transcript_table[\"transcript_start\"]).to_list()\n","        n=0\n","        for i in range(len(loc)):\n","            item=loc[i]\n","            if (wt_seq[-item]==transcript_table[\"V6.x\"].to_list()[i])==False:\n","                n+=1\n","            #else:\n","               # wt_seq[-item]=transcript_table[\"V7.x\"].to_list()[i] \n","        print(\"Mismatch between reference and ensembl: \"+str(n))\n","        wt_seq=\"\".join(wt_seq)\n","        wt_seq=compliment(wt_seq)\n","    else:\n","        loc=(transcript_table[\"V2.x\"]-transcript_table[\"transcript_start\"]).to_list()\n","        wt_seq=list(wt_seq.strip())\n","        n=0\n","        for i in range(len(loc)):\n","            item=loc[i]\n","            if (wt_seq[item]==transcript_table[\"V6.x\"].to_list()[i])==False:\n","                n+=1\n","           # else:\n","             #   wt_seq[item]=transcript_table[\"V7.x\"].to_list()[i]\n","        print(\"Mismatch between reference and ensembl: \"+str(n))\n","        wt_seq=\"\".join(wt_seq)\n","    return wt_seq"]},{"cell_type":"code","execution_count":null,"id":"30d92e55","metadata":{"id":"30d92e55"},"outputs":[],"source":["# def mutants(transcript_table,id,wt_seq):\n","# #introduce mutation\n","#     transcript_table=transcript_table[transcript_table[\"transcript_id\"].apply(lambda x: x.startswith(id))]\n","#     loc=(transcript_table[\"V2.x\"]-transcript_table[\"transcript_start\"]).to_list()\n","#     wt_seq=list(wt_seq.strip())\n","#     n=0\n","#     for i in range(len(loc)):\n","#         item=loc[i]\n","#         if (wt_seq[item]==transcript_table[\"V6.x\"].to_list()[i])==False:\n","#             n+=1\n","#        # else:\n","#          #   wt_seq[item]=transcript_table[\"V7.x\"].to_list()[i]\n","#     print(\"Mismatch between reference and ensembl: \"+str(n))\n","#     return \"\".join(wt_seq)"]},{"cell_type":"code","execution_count":null,"id":"5d2b4a62","metadata":{"id":"5d2b4a62"},"outputs":[],"source":["def exons(table,wt_seq):\n","    relative_start=[]\n","    start=table[\"Start\"].to_list()[0]\n","    for item in table[\"Start\"].to_list():\n","        relative_start.append(np.abs(item-start))\n","    relative_end=[]\n","    for item in table[\"End\"].to_list():\n","        relative_end.append(np.abs(item-start)+1)\n","    final_seq=\"\"\n","    for item in zip(relative_start,relative_end):\n","        final_seq+=wt_seq[item[0]:item[1]]\n","    return final_seq"]},{"cell_type":"code","execution_count":null,"id":"f87e4fdc","metadata":{"id":"f87e4fdc"},"outputs":[],"source":["def translate(seq,seq_dict): \n","    table = {\n","        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n","        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n","        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n","        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                \n","        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n","        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n","        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n","        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n","        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n","        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n","        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n","        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n","        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n","        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n","        'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n","        'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',\n","    }\n","    loc=seq.find(\"ATG\")\n","    seq=seq[loc:]\n","    protein =\"\"\n","    while (len(protein)!=len(seq_dict[\"amino\"]) and (\"ATG\" in seq)):\n","        protein=\"\"\n","        for i in range(0, len(seq), 3):\n","            codon = seq[i:i + 3]\n","            if table[codon]!=\"_\":\n","                protein+= table[codon]\n","            else:\n","                break\n","        if len(protein)!=len(seq_dict[\"amino\"]):\n","            seq=seq[1:]\n","            if \"ATG\" in seq:\n","                loc=seq.find(\"ATG\")\n","                seq=seq[loc:]\n","            else:\n","                break\n","        else:\n","            break\n","    return protein"]},{"cell_type":"code","execution_count":null,"id":"92d97567","metadata":{"id":"92d97567"},"outputs":[],"source":["def workflow(filename,transcript_table,seq_dict,id):\n","    table=table_info(filename)\n","    wt_seq=seq_dict[\"genomic\"]\n","    flag=anti(table)\n","    mutant_seq=mutants(transcript_table,id,wt_seq,flag)\n","    print(\"mutant:\" + str(mutant_seq==seq_dict[\"genomic\"]) )\n","    exon=exons(table,mutant_seq)\n","    print(\"Exon:\" + str(exon==seq_dict[\"exon_only\"]))\n","    protein=translate(exon,seq_dict)\n","    print(\"Protein: \"+ str(protein==seq_dict[\"amino\"]))\n","    return protein,exon"]},{"cell_type":"code","execution_count":null,"id":"be72835b","metadata":{"id":"be72835b"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"25e1db69","metadata":{"id":"25e1db69"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"ff594e46","metadata":{"id":"ff594e46"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"c40467d1","metadata":{"id":"c40467d1"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"7bd4f516","metadata":{"id":"7bd4f516"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"e3f3e99f","metadata":{"id":"e3f3e99f"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"44ba8df2","metadata":{"id":"44ba8df2"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"4bb63362","metadata":{"id":"4bb63362"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"251c3c0e","metadata":{"id":"251c3c0e"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"4229f5e5","metadata":{"id":"4229f5e5"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"bf480397","metadata":{"id":"bf480397"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"8c66e46c","metadata":{"id":"8c66e46c"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"fbcddfe3","metadata":{"id":"fbcddfe3"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"5c7d1d47","metadata":{"id":"5c7d1d47"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"9be03d06","metadata":{"id":"9be03d06"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"950b6d1b","metadata":{"id":"950b6d1b"},"outputs":[],"source":[]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.8"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":5}